xtask/src/quirk.rs

```diff
@@ -0,0 +1,148 @@
+use strum_macros::{Display, EnumIter};
+
+#[derive(Debug, Display, Copy, Clone, PartialEq, EnumIter)]
+pub enum LauncherMode {
+    Virtualized,
+}
+
+impl LauncherMode {
+    /// Get the crate name of respective enclave binary variant
+    pub fn enclave_crate_name(&self) -> &'static str {
+        match self {
+            LauncherMode::Virtualized => "quirk_echo_enclave_app",
+        }
+    }
+
+    /// Get the path to the respective enclave binary variant that should be launched
+    pub fn enclave_crate_path(&self) -> String {
+        format!("./{}", self.enclave_crate_name())
+    }
+
+    /// Get the path to the respective enclave binary variant that should be launched
+    pub fn enclave_binary_path(&self) -> String {
+        match self {
+            LauncherMode::Virtualized => format!(
+                "{}/target/x86_64-unknown-none/debug/{}",
+                self.enclave_crate_path(),
+                self.enclave_crate_name()
+            ),
+        }
+    }
+
+    /// Get the subcommand for launching in this mode
+    pub fn variant_subcommand(&self) -> Vec<String> {
+        match self {
+            LauncherMode::Virtualized => vec![
+                "virtualized".to_string(),
+                format!(
+                    "--enclave-binary={}",
+                    "./oak_restricted_kernel_bin/target/x86_64-unknown-none/debug/oak_restricted_kernel_bin"
+                ),
+                format!("--vmm-binary={}", "/usr/bin/qemu-system-x86_64"),
+                format!("--app-binary={}", &self.enclave_binary_path()),
+                format!(
+                    "--bios-binary={}",
+                    "./stage0/target/x86_64-unknown-none/release/oak_stage0.bin"
+                ),
+            ],
+        }
+    }
+}
+
+fn run_variant(variant: LauncherMode) -> Step {
+    let mut steps = vec![build_binary(
+        "build Quirk Echo enclave binary",
+        &variant.enclave_crate_path(),
+    )];
+    // If we want to run in an VMM, we need three binaries:
+    // 1. the stage0 BIOS image,
+    // 2. the kernel binary,
+    // 3. the actual Quirk Echo enclave application.
+    // (1) and (2) are needed to start the VMM, and the kernel expects to read (3) as the very first
+    // thing over the communication channel.
+    steps.extend(match variant {
+        LauncherMode::Virtualized => vec![
+            build_stage0(),
+            build_binary(
+                "build Restricted Kernel binary",
+                "oak_restricted_kernel_bin",
+            ),
+        ],
+    });
+    steps.push(Step::WithBackground {
+        name: "background launcher".to_string(),
+        background: run_launcher(variant),
+        foreground: Box::new(Step::Single {
+            name: "waiting for 10 seconds".to_string(),
+            command: Cmd::new("sleep", ["10"]),
+        }),
+    });
+    Step::Multiple {
+        name: format!("run {} variant", variant),
+        steps,
+    }
+}
+
+fn build_stage0() -> Step {
```

Comment:
This and various other functions seem identical to the existing ones. Please reuse them as much as possible, and feel free to keep the implementations that are currently different. In the future we can modularize the existing code to make it simpler to share even those parts by parametrizing them by binary names etc..

Reply:
Reused shared code from launcher in quirk xtask.